### ISSUE TITLE
Discrepancy in test statistics from cacluators

### ISSUE THREAD

**Issue by matthewfeickert** [scikit-hep/pyhf#1792]
### Summary

There is a large discrepancy in the value of the discovery test statistic if it is generated via an Asymptotic based or toy based calculator.

Related Issues:
* https://github.com/scikit-hep/pyhf/issues/1587

### OS / Environment

```console
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

### Steps to Reproduce

```python (paste below)
import pyhf


def discovery_from_test_stat(model, data):
    init_pars = model.config.suggested_init()
    par_bounds = model.config.suggested_bounds()
    fixed_params = model.config.suggested_fixed()
    return pyhf.infer.test_statistics.q0(
        0.0,
        data,
        model,
        init_pars,
        par_bounds,
        fixed_params,
    )


def discovery_from_calculator(model, data, calc_type="asymptotics", **kwargs):
    calculator = pyhf.infer.utils.create_calculator(
        calc_type, data, model, test_stat="q0", **kwargs
    )
    return calculator.teststatistic(0.0)


def discovery_from_asymptotic_calculator(model, data):
    calculator = pyhf.infer.calculators.AsymptoticCalculator(
        data, model, test_stat="q0"
    )
    return calculator.teststatistic(0.0)


if __name__ == "__main__":
    model = pyhf.simplemodels.uncorrelated_background([25], [2500], [2.5])
    data = [2600] + model.config.auxdata

    discovery_test_stat = discovery_from_test_stat(model, data)
    print(f"Discovery test stat from infer.test_statistics.q0: {discovery_test_stat}")

    discovery_test_stat = discovery_from_calculator(model, data, calc_type="toybased")
    print(
        f"Discovery test stat from toy based infer.utils.create_calculator: {discovery_test_stat}"
    )

    discovery_test_stat = discovery_from_calculator(model, data)
    print(
        f"Discovery test stat from asymptotics infer.utils.create_calculator: {discovery_test_stat}"
    )

    discovery_test_stat = discovery_from_asymptotic_calculator(model, data)
    print(
        f"Discovery test stat from infer.calculators.AsymptoticCalculator: {discovery_test_stat}"
    )
```

```console (paste below)

```

### Expected Results

The discovery test statistic would be the same regardless of calculator type.

### Actual Results

```console
$ python discovery_bug.py 
Discovery test stat from infer.test_statistics.q0: 3.9377335959507036
Discovery test stat from toy based infer.utils.create_calculator: 3.9377335959507036
Discovery test stat from asymptotics infer.utils.create_calculator: 1.485845489706193
Discovery test stat from infer.calculators.AsymptoticCalculator: 1.485845489706193
```

### pyhf Version

```console
9fd99be886349a90e927672e950cc233fad0916c on master
```

### Code of Conduct

- [X] I agree to follow the Code of Conduct

**Comment by matthewfeickert** [comment]
> are you sure you have enough toys to evaluate this (several 10k)?

I need to revisit how all of this is implemented, but I'm pretty sure there's a bug somewhere or an API that is very unclear as for

https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/calculators.py#L663

the 

https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/calculators.py#L922

added back in PR #790 isn't using toy information at all

https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/calculators.py#L951-L960

(Maybe I'm missing something obvious from the past, but at the moment I'm confused).

**Comment by alexander-held** [comment]
The last two numbers scale with the number of signal events, so these are something else.

**Comment by matthewfeickert** [comment]
> The last two numbers scale with the number of signal events, so these are something else.

yeah

**Comment by alexander-held** [comment]
At the moment this doesn't run any toys as far as I can tell, so I'm assuming this is a question about the calculators, and not about asymptotic vs toy agreement in general?

**Comment by matthewfeickert** [comment]
> At the moment this doesn't run any toys as far as I can tell, so I'm assuming this is a question about the calculators, and not about asymptotic vs toy agreement in general?

Yes, I didn't phrase the original text clearly as I was dumping things in for myself to clarify later.

**Comment by matthewfeickert** [comment]
Okay @kratsg has pointed out that the behavior of the calculator APIs is (known to be) not consistent across the asymptotic and toy based as asymptotic is returning p-values

https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/__init__.py#L169-L178

while the toy based is returning q test stats.

**Comment by kratsg** [comment]
> APIs is (known to be) not consistent across the asymptotic and toy based as asymptotic is returning p-values

well, it's consistent here, it's that the `calc.teststatistic(test_poi)` has a different meaning which is translated by the respective calculator's `calc.pvalues(teststat, ...)` call if that makes sense. The API is the same, but there's a few hoops/hurdles/threading to make it work.

**Comment by matthewfeickert** [comment]
> well, it's consistent here, it's that the `calc.teststatistic(test_poi)` has a different meaning which is translated by the respective calculator's `calc.pvalues(teststat, ...)` call if that makes sense. The API is the same, but there's a few hoops/hurdles/threading to make it work.

While all very true, a user should rightly complain that the public API is too confusing as is (given the current documentation).

**Comment by matthewfeickert** [comment]
My main complaint at the moment is that while we make it clear that the asymptotic test stats are in `-^mu/sigma` space

https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/calculators.py#L85-L89

we don't make this clear again in the calculator `teststatistic` API for the asymptotics calculator 

https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/calculators.py#L330-L332

or in the toy calculator (that it is different)


https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/calculators.py#L922-L924

Also we could mention in `EmpiricalDistribution` that the test stats are distributed in different spaces as well

https://github.com/scikit-hep/pyhf/blob/9fd99be886349a90e927672e950cc233fad0916c/src/pyhf/infer/calculators.py#L526-L532

So this really is a documentation issue, but a pretty big one in my mind.

I forgot this, and if I've written some of this code and can make this mistake when tired then I think a user definitely will.

**Comment by kratsg** [comment]
> My main complaint at the moment is that while we make it clear that the asymptotic test stats are in `-^mu/sigma` space

this part I think is what confuses me, so if you have a better grasp, it would be great to clarify this for the upcoming release.

**Comment by matthewfeickert** [comment]
This all came up as I was trying to take a stab at https://github.com/scikit-hep/pyhf/issues/1712 and was trying to figure out to have things work for either calculator type.
